### DISCUSSION TITLE
Listening for changes in preview buffer

### DISCUSSION THREAD

**Discussion by ollien** [vscode-neovim/vscode-neovim#1992] [Development]
Following @xiyaowong's awesome work in getting `inccommand=split` working with our completion system (https://github.com/vscode-neovim/vscode-neovim/pull/1985), I was curious to see if I could iterate on it by listening for changes to the preview buffer itself, rather than having to wait for a small delay after `CmdlineChanged`. I was hoping that `nvim_buf_attach` would allow me to capture these changes, but alas, no dice.

A super quick prototype looked like this (you can source this in an nvim window with `icm=split` set). You'll notice if you look at `:messages`, `on_line` is never fired. Is there something special about the preview buffer that excludes it from working with `nvim_buf_attach`'s callbacks?

```lua
local api, fn = vim.api, vim.fn

local function find_preview_buf()
  local preview_buf
  local bufs = api.nvim_list_bufs()
  for _, buf in ipairs(bufs) do
    if
      vim.bo[buf].buftype == "nofile"
      and fn.bufname(buf) == "[Preview]"
    then
      preview_buf = buf
      break
    end
  end

  return preview_buf
end

local split_listener_registered = false
local function listen_for_split_completion()
  if split_listener_registered then
    return
  end

  local preview_buf = find_preview_buf()
  if not preview_buf then
    return
  end

  attached = api.nvim_buf_attach(preview_buf, false, {
    on_lines = function(_, buf, changedtick, firstline, lastline, new_lastline, bytecount)
      print(string.format("Buffer %d changed: %d-%d to %d", buf, firstline, lastline, new_lastline))
    end,
    preview = true,
  })
  split_listener_registered = true
end

api.nvim_create_autocmd({ "CmdlineEnter" }, {
  callback = function()
    split_listener_registered = false
  end,
})

api.nvim_create_autocmd({ "CmdlineChanged" }, {
  callback = listen_for_split_completion,
})
```

**Comment by xiyaowong** [comment]
It is a special window and a special buffer, where many events are absent. Additionally, this has complicated the logic without bringing any benefits. Consider the following:

1. The buffer no longer exists.
2. Avoid conflicts with wildmenu_show.
3. Avoid being affected by wildmenu_hide.

**Reply by ollien** [reply]
Nope, understood. Like I said, this was a small thing I threw together to better understand what I could do. There is a benefit to this approach though, if it worked :) there is a delay between typing something and it appearing in the wld_menu, which was a bit frustrating.
